x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/integration/MlDistributedFailureIT.java

```diff
@@ -245,19 +245,24 @@ public void testCloseUnassignedFailedJobAndStopUnassignedStoppingDatafeed() thro
         // stopping.
         PersistentTasksCustomMetaData tasks = clusterService().state().getMetaData().custom(PersistentTasksCustomMetaData.TYPE);
         PersistentTasksCustomMetaData.PersistentTask<?> task = MlTasks.getDatafeedTask(datafeedId, tasks);
-        UpdatePersistentTaskStatusAction.Request updatePersistentTaskStatusRequest =
-            new UpdatePersistentTaskStatusAction.Request(task.getId(), task.getAllocationId(), DatafeedState.STOPPING);
-        PersistentTaskResponse updatePersistentTaskStatusResponse =
-            client().execute(UpdatePersistentTaskStatusAction.INSTANCE, updatePersistentTaskStatusRequest).actionGet();
-        assertNotNull(updatePersistentTaskStatusResponse.getTask());
 
-        // Confirm the datafeed state is now stopping - this may take a while to update in cluster state
-        assertBusy(() -> {
-            GetDatafeedsStatsAction.Request datafeedStatsRequest = new GetDatafeedsStatsAction.Request(datafeedId);
-            GetDatafeedsStatsAction.Response datafeedStatsResponse =
-                client().execute(GetDatafeedsStatsAction.INSTANCE, datafeedStatsRequest).actionGet();
-            assertEquals(DatafeedState.STOPPING, datafeedStatsResponse.getResponse().results().get(0).getDatafeedState());
-        });
+        // It is possible that the datafeed has already detected the job failure and
+        // terminated itself. In this happens there is no task
+        if (task != null) {
```

Comment:
The problem with this is that if `task == null` then the test isn't doing what it's supposed to be doing, which is testing that you can stop an unassigned `stopping` datafeed.

Since this happens so rarely we could put the whole test in a loop and retry a few times if this situation occurs.

But if you don't want to do that then it would be better to use `assumeFalse("Test setup did not create the required conditions", task == null);` because then at least the test will be reported as having been skipped rather than silently succeeding when it didn't test what it was supposed to test.

Reply:
I think assumeFalse is best, retry adds complexity to an already complicated piece of code. The test could fail for any number of reasons some including general infrastructure problems retrying would add noise. 

Looking at the test, in the setup it waits for the datafeed to complete before closing the job which should have prevented flush happening after the job failed.

https://github.com/elastic/elasticsearch/blob/071b60db2941e54c19446403d4fcd49d9d3a4f9f/x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/integration/MlDistributedFailureIT.java#L215

`waitForDatafeed` waits for the data counts to be indexed which the datafeed does asynchronously before flushing the job. It is amazing that the race between the writing the datacounts and calling flush is sometimes lost to a thread doing an `assertBusy` searching the index for the new datacounts and then failing the job but that is whats happening here. This is so unlikely (the first time its been reported) it's not worth rewriting the test.